python/pip.bzl

```diff
@@ -13,161 +13,51 @@
-pip using the same interpreter as your toolchain. If set, takes precedence over
-python_interpreter.
-"""),
-        "requirements": attr.label(
-            mandatory = True,
-            allow_single_file = True,
-            doc = "The label of the requirements.txt file.",
-        ),
-        "timeout": attr.int(
-            default = 600,
-            doc = "Timeout (in seconds) for repository fetch."
-        ),
-        "_script": attr.label(
-            executable = True,
-            default = Label("//tools:piptool.par"),
-            cfg = "host",
-        ),
-    },
-    implementation = _pip_import_impl,
-    doc = """A rule for importing `requirements.txt` dependencies into Bazel.
-
-This rule imports a `requirements.txt` file and generates a new
-`requirements.bzl` file.  This is used via the `WORKSPACE` pattern:
-
-```python
-pip_import(
-    name = "foo",
-    requirements = ":requirements.txt",
-)
-load("@foo//:requirements.bzl", "pip_install")
-pip_install()
-```
-
-You can then reference imported dependencies from your `BUILD` file with:
-
-```python
-load("@foo//:requirements.bzl", "requirement")
-py_library(
-    name = "bar",
-    ...
-    deps = [
-       "//my/other:dep",
-       requirement("futures"),
-       requirement("mock"),
-    ],
-)
-```
-
-Or alternatively:
-```python
-load("@foo//:requirements.bzl", "all_requirements")
-py_binary(
-    name = "baz",
-    ...
-    deps = [
-       ":foo",
-    ] + all_requirements,
-)
-```
-""",
-)
-
-# We don't provide a `pip2_import` that would use the `python2` system command
-# because this command does not exist on all platforms. On most (but not all)
-# systems, `python` means Python 2 anyway. See also #258.
-
-def pip3_import(**kwargs):
-    """A wrapper around pip_import that uses the `python3` system command.
-
-    Use this for requirements of PY3 programs.
+load("//python/pip_install:pip_repository.bzl", "pip_repository")
+load("//python/pip_install:repositories.bzl", "pip_install_dependencies")
+
+def pip_install(requirements, name = "pip", **kwargs):
+    """Imports a `requirements.txt` file and generates a new `requirements.bzl` file.
+
+    This is used via the `WORKSPACE` pattern:
+
+    ```python
+    pip_install(
+        requirements = ":requirements.txt",
+    )
+    ```
+
+    You can then reference imported dependencies from your `BUILD` file with:
+
+    ```python
+    load("@pip//:requirements.bzl", "requirement")
+    py_library(
+        name = "bar",
+        ...
+        deps = [
+           "//my/other:dep",
+           requirement("requests"),
+           requirement("numpy"),
+        ],
+    )
+    ```
+
+    Args:
+      requirements: A 'requirements.txt' pip requirements file.
+      name: A unique name for the created external repository (default 'pip').
+      **kwargs: Keyword arguments passed directly to the `pip_repository` repository rule.
     """
-    pip_import(python_interpreter = "python3", **kwargs)
-
-def pip_repositories():
```

Comment:
Doing this is unfortunately not backwards compatible with `rules_docker`: https://github.com/bazelbuild/rules_docker/blob/master/repositories/py_repositories.bzl#L34

We can leave it in and/or submit a change to `rules_docker` to remove what's currently just a "placeholder"

Reply:
Example of the failure: 

```
ERROR: /private/var/tmp/_bazel_jbelotti/a8a1f81a1422a72c79c6c81641a8c816/external/io_bazel_rules_docker/repositories/py_repositories.bzl:22:1: file '@rules_python//python:pip.bzl' does not contain symbol 'pip_repositories'
ERROR: error loading package '': in /private/var/tmp/_bazel_jbelotti/a8a1f81a1422a72c79c6c81641a8c816/external/io_bazel_rules_docker/repositories/deps.bzl: Extension file 'repositories/py_repositories.bzl' has errors
```

Reply:
ah yeah, that one always gets us, I'll add back the function as

```
def pip_repositories():
    # buildifier: disable=print
    print("DEPRECATED: the pip_repositories rule has been replaced with pip_install, please see release notes")
```

Reply:
I should've gone further down with my comment, as right after you hit a `pip_import` usage which has the same problem: https://github.com/bazelbuild/rules_docker/blob/master/repositories/py_repositories.bzl#L36

The implication of this is that upgrading to `0.1.*` will make `rules_python` incompatible with any existing repositories that a `WORKSPACE` pulls and that use `pip_import`. Those repos would need to get a PR and then ship a release that makes itself compatible with `0.1.*`. Some of those repos would include: 

* https://github.com/etsy/rules_grafana/blob/ff452327df3dcec541e861bb0fe31114c4294bec/grafana/workspace.bzl#L2
* https://github.com/graknlabs/bazel-distribution/blob/b6ae4297ab66484107f9469d88a4e5d047b36c53/pip/deps.bzl
* https://github.com/stackb/rules_proto/blob/1d6b84118399828511faeecc145d399c1e7bdee2/WORKSPACE
* `rules_docker` (already noted here)

Reply:
those `pip_import` rules will also print the warning, and yeah they'll need an update to get past the breaking change. We can send those PRs out - is that your proposal? Obviously end-users need to wait until all their deps are compatible before they can update (we've had the same thing with rules_docker in some breaking changes for rules_nodejs)

OR are you suggesting we do something more to adapt the existing API for these users to make it less breaking?

Reply:
> We can send those PRs out - is that your proposal? 

Yeah I think we'd be better off being proactive about it.

Reply:
>  are you suggesting we do something more to adapt the existing API for these users to make it less breaking?

I thought about it. These Bazel repo rules are awkward to version as you can't declare your repo rule deps, things just break at sync/build time.